answer.go

```diff
@@ -193,19 +194,19 @@ func (p *Promise) resolve(r Ptr, e error) {
 		if p.ongoingCalls > 0 {
 			p.callsStopped = make(chan struct{})
 		}
-		p.mu.Unlock()
-		res := resolution{p.method, r, e}
-		for path, row := range p.clients {
-			t := path.transform()
-			for i := range row {
-				row[i].promise.Fulfill(res.client(t))
-				row[i].promise = nil
+		syncutil.Without(&p.mu, func() {
```

Comment:
Just a small idea/suggestion:

I have a small library called [`github.com/lthibault/util/sync`](https://github.com/lthibault/util/sync) into which these functions would fit quite nicely.  By happy coincidence, the package name is also `syncutil`.  It could help declutter things a bit if we were to move it there and import the library.

Reply:
Makes sense to me. I'm going to hit merge; feel free to move these and submit a pr.

It might also make sense to pull some other random packages out of `internal`, e.g. `mpsc`, `chanmutex`...

I am vaguely bothered by the fact that your package path doesn't agree with the package name, but meh.

Reply:
>I am vaguely bothered by the fact that your package path doesn't agree with the package name, but meh.

I used to be the same way, but the `import project "github.com/user/project/pkg"` convention is quite popular, so I don't think it's "wrong" in Go.

The alternative was a name collision with the standard library `sync`.

Anyway 👍 .  Sounds like a plan.  Merge it!